www/main.js

```diff
@@ -11,6 +11,64 @@ function loadDefaultData() {
 }
 loadDefaultData()
 
+// Create and configure a new web worker to run python code
+function createCodeWorker() {
+  const codeWorker = new Worker('/worker.js')
+
+  codeWorker.addEventListener('message', function (msg) {
+    console.log('Message received')
+
+    if (msg.data.type === 'result') {
+      document.querySelector('#code-output').innerHTML = msg.data.result
+      runButton.disabled = false
+      endButton.disabled = true
+    }
+  })
+
+  return codeWorker
+}
+
+// Run python code in web worker and deal with run button
+let codeWorker = createCodeWorker()
+// let workerIsDead = false
+
+// get HTML elements
+const runButton = document.querySelector('#run-button')
+const endButton = document.querySelector('#end-button')
+const timeDisplayP = document.querySelector('#time-displayed')
+const saveOutput = document.querySelector('#save-output')
+
+// Run code when button pressed.
+runButton.addEventListener('click', function () {
+  //
+  // if (workerIsDead) {
+  //   codeWorker = createCodeWorker()
+  // }
+  const studentCode = document.querySelector('#code-area').value
+  runButton.disabled = true
+  endButton.disabled = false
+
+  console.log('Posting message')
+  codeWorker.postMessage({
+    type: 'run',
+    language: 'python',
+    code: studentCode,
+  })
+})
+
+endButton.addEventListener('click', function () {
+  codeWorker.terminate()
+  codeWorker = createCodeWorker() // send in the next worker
+
+  runButton.disabled = false
+  endButton.disabled = true
+})
+
+saveOutput.addEventListener('click', function () {
+  document.querySelector('#output-text').value =
+    document.querySelector('#code-output').innerHTML
+})
+
```

Comment:
It seems like this code should be shared somehow between the main page and the student page rather than being duplicated in two separate files.

Reply:
I agree that this is something that should be done but I don't think there is time to do that until the next itteration